moshi/src/main/java/com/squareup/moshi/Moshi.kt

```diff
@@ -0,0 +1,388 @@
+      jsonAdapter: JsonAdapter<T>
+    ): Builder = apply {
+      add(newAdapterFactory(type, annotation, jsonAdapter))
+    }
+
+    public fun add(factory: JsonAdapter.Factory): Builder = apply {
+      factories.add(lastOffset++, factory)
+    }
+
+    public fun add(adapter: Any): Builder = apply {
+      add(AdapterMethodsFactory.get(adapter))
+    }
+
+    @Suppress("unused")
+    public fun <T> addLast(type: Type, jsonAdapter: JsonAdapter<T>): Builder = apply {
+      addLast(
+        newAdapterFactory(
+          type,
+          jsonAdapter
+        )
+      )
+    }
+
+    @Suppress("unused")
+    public fun <T> addLast(
+      type: Type,
+      annotation: Class<out Annotation>,
+      jsonAdapter: JsonAdapter<T>
+    ): Builder = apply {
+      addLast(newAdapterFactory(type, annotation, jsonAdapter))
+    }
+
+    public fun addLast(factory: JsonAdapter.Factory): Builder = apply {
+      factories.add(factory)
+    }
+
+    @Suppress("unused")
+    public fun addLast(adapter: Any): Builder = apply {
+      addLast(AdapterMethodsFactory.get(adapter))
+    }
+
+    @CheckReturnValue
+    public fun build(): Moshi = Moshi(this)
+  }
+
+  /**
+   * A possibly-reentrant chain of lookups for JSON adapters.
+   *
+   * We keep track of the current stack of lookups: we may start by looking up the JSON adapter
+   * for Employee, re-enter looking for the JSON adapter of HomeAddress, and re-enter again looking
+   * up the JSON adapter of PostalCode. If any of these lookups fail we can provide a stack trace
+   * with all of the lookups.
+   *
+   * Sometimes a JSON adapter factory depends on its own product; either directly or indirectly.
+   * To make this work, we offer a JSON adapter stub while the final adapter is being computed. When
+   * it is ready, we wire the stub to that finished adapter. This is necessary in self-referential
+   * object models, such as an `Employee` class that has a `List<Employee>` field for an
+   * organization's management hierarchy.
+   *
+   * This class defers putting any JSON adapters in the cache until the topmost JSON adapter has
+   * successfully been computed. That way we don't pollute the cache with incomplete stubs, or
+   * adapters that may transitively depend on incomplete stubs.
+   */
+  internal inner class LookupChain {
+    private val callLookups: MutableList<Lookup<*>> = ArrayList()
+    private val stack = ArrayDeque<Lookup<*>>()
+    private var exceptionAnnotated = false
+
+    /**
+     * Returns a JSON adapter that was already created for this call, or null if this is the first
+     * time in this call that the cache key has been requested in this call. This may return a
+     * lookup that isn't yet ready if this lookup is reentrant.
+     */
+    fun <T> push(type: Type, fieldName: String?, cacheKey: Any): JsonAdapter<T>? {
+      // Try to find a lookup with the same key for the same call.
+      var i = 0
+      val size = callLookups.size
+      while (i < size) {
```

Comment:
For seemed to be a better fit here..

Reply:
You can't write for loops in Kotlin with multiple variables